sender/nats.go

```diff
@@ -0,0 +1,171 @@
+/*
+ * skogul, nats producer/sender
+ *
+ * Author(s):
+ *  - Niklas Holmstedt <n.holmstedt@gmail.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301  USA
+ */
+
+ package sender
+
+ import (
+	"fmt"
+	"github.com/nats-io/nats.go"
+	"github.com/telenornms/skogul"
+	"github.com/telenornms/skogul/encoder"
+	"sync"
+	"crypto/tls"
+)
+
+var natsLog = skogul.Logger("sender", "nats")
+/*
+Nats.io sender. A small Nats, non-jetstream, publisher with:
+
+- Authentication: Username/Password, TLS
+- Authorization: Username/Password, UserCredentials/JWT
+
+*/
+
+type Nats struct {
+	Servers		string   `doc:"Comma separated list of nats URLs"`
+	Subject		string   `doc:"Subject to publish messages on"`
+	Name		string	 `doc:"Client name"`
+	Username	string   `doc:"Client username"`
+	Password	string	 `doc:"Client password"`
+	TLSClientKey    string   `doc:"TLS client key file path"`
+	TLSClientCert   string   `doc:"TLS client cert file path"`
+	TLSCACert	string   `doc:"CA cert file path"`
+	UserCreds       string   `doc:"Nats credentials file path"`
+	NKeyFile        string   `doc:"Nats nkey file path"`
+	Insecure	bool	 `doc:"TLS InsecureSkipVerify"`
+	Encoder		skogul.EncoderRef
+	o		*[]nats.Option
+	nc		*nats.Conn
+	once		sync.Once
+}
+
+// Verify configuration
+func (n *Nats) Verify() error {
+        if n.Subject == "" {
+                return skogul.MissingArgument("Subject")
+        }
+
+        return nil
+}
+
+
+func (n *Nats) init() {
+
+	if n.Encoder.Name == "" {
+		n.Encoder.E = encoder.JSON{}
+	}
+
+	if n.Name == "" {
+		n.Name = "skogul"
+	}
+	n.o = &[]nats.Option{nats.Name(n.Name)}
+
+	if n.Servers == "" {
+		n.Servers = nats.DefaultURL
+	}
+
+	//User Credentials, use either.
+	if n.UserCreds != "" && n.NKeyFile != "" {
+		natsLog.Fatal("Please configure usercreds or nkeyfile.")
```

Comment:
Should be part of Verify so it fails on skogul -show, not only on attempted start.